### PR TITLE
PWX-29354: Update StorageClass on PV post PVC migration

### PR DESCRIPTION
**Pull request by pp511**
**What type of PR is this?**
>improvement

**What this PR does / why we need it**:
https://portworx.atlassian.net/browse/PWX-29354 
https://portworx.atlassian.net/browse/CEE-804
Some operators require StorageClass to be present on PV to be able to consume the resources.

**Does this PR change a user-facing CRD or CLI?**:
no

**Is a release note needed?**:
yes
```
* Update StorageClass on PersistentVolumes during migration.
```

**Does this change need to be cherry-picked to a release branch?**:
yes 23.3

**Jenkins Job for migration regressions**
https://jenkins.pwx.dev.purestorage.com/job/Stork/view/Stork%2023.2-dev%20%20C7/job/23-2-migration-k8s-1-24-0/384/parameters/

### CODE REVIEW

File: pkg/migration/controllers/migration.go [hunk 1]
```diff
@@ -1564,6 +1575,82 @@ func (m *MigrationController) getParsedLabels(
...
+				sc, err := storkcache.Instance().GetStorageClass(scName)
+				if err != nil {
+					return err
+				}
+				sc.UID = ""
+				sc.ResourceVersion = ""
+				log.MigrationLog(migration).Infof("Applying %v %v", sc.Kind, sc.Name)
+				for retries := 0; retries < maxApplyRetries; retries++ {
+					_, err = remoteClient.adminClient.StorageV1().StorageClasses().Create(context.TODO(), sc, metav1.CreateOptions{})
+					if err == nil {
```
Comment (diptiranjanpx) [inline pkg/migration/controllers/migration.go:1624]: Can we add err == nil || errors.IsAlreadyExists(err) ?

Comment (diptiranjanpx) [reply pkg/migration/controllers/migration.go:1624]: Also what will happen if there exists another storageclass with same name but different parameters or even provisioners? this case is very unlikely though.

**Review by diptiranjanpx — APPROVED** [review]

File: pkg/migration/controllers/migration.go (hunk 1 above)
Comment (pp511) [reply pkg/migration/controllers/migration.go:1624]: 1. We are only creating the SC if it doens't exist. So I think we will never get "IsAlreadyExists". However, it still is logical to handle this condition. Updated as you mentioned.
2. We don't want to migrate SC if it already exists on destination. So if Sc exists with different params, we will not migrate the SC. We would still update the SC name to the PV.

File: pkg/migration/controllers/migration.go [hunk 2]
```diff
@@ -1564,6 +1575,82 @@ func (m *MigrationController) getParsedLabels(
...
+				sc, err := storkcache.Instance().GetStorageClass(scName)
+				if err != nil {
+					return err
+				}
+				sc.UID = ""
+				sc.ResourceVersion = ""
+				log.MigrationLog(migration).Infof("Applying %v %v", sc.Kind, sc.Name)
+				for retries := 0; retries < maxApplyRetries; retries++ {
+					_, err = remoteClient.adminClient.StorageV1().StorageClasses().Create(context.TODO(), sc, metav1.CreateOptions{})
+					if err == nil || errors.IsAlreadyExists(err) {
```
Comment (diptiranjanpx) [inline pkg/migration/controllers/migration.go:1624]: We may hit the nil pointer derereference. 
Lets make it like 
if err == nil || ( err != nil && errors.IsAlreadyExists(err))

Comment (pp511) [reply pkg/migration/controllers/migration.go:1624]: rhs of || will only be computed if err is not nil. So we don't need to check err != nil  condition explicitly.

Comment (diptiranjanpx) [reply pkg/migration/controllers/migration.go:1624]: right. Ignore the comment.

**Pull request merged**